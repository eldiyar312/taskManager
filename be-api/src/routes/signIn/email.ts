import { Handler, SignInEmailPost } from 'src/_generated';
import { authWithEmail } from 'src/middlewares/passport';
import { User } from 'src/models/user';
import { userToProfile } from 'src/storage/profile';
import { config } from 'src/utils/navigation';

export const options = config('Sign In with email and password', {
  tag: 'access',
  roles: ['guest'],
  schema: {
    body: {
      type: 'object',
      required: ['email', 'password'],
      properties: {
        email: { type: 'string', format: 'email' },
        password: { type: 'string' },
      },
      additionalProperties: false,
    },
    response: {
      200: { $ref: 'models#/properties/profile' },
      401: { $ref: 'errors#/properties/simple' },
    },
  },
  preHandler: [authWithEmail],
});

export const handler: Handler<SignInEmailPost> = (request) =>
  userToProfile(request.user as User);
