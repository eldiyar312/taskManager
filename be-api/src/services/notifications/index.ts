export { email } from 'src/services/notifications/email';
