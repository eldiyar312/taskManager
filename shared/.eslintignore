# dependencies
build/
node_modules/
/.pnp
.pnp.js
*.d.ts

# misc
.DS_Store

yarn-debug.log*
yarn-error.log*

typings
