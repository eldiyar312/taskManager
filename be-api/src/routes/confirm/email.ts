import dayjs from 'dayjs';

import { ERRORS } from 'shared';

import { ConfirmEmailPost, Handler } from 'src/_generated';
import { signIn } from 'src/middlewares/passport';
import { User } from 'src/models/user';
import { userToProfile } from 'src/storage/profile';
import {
  confirmUserEmail,
  findUserByEmailConfirmationCode,
} from 'src/storage/users';
import { config } from 'src/utils/navigation';

export const options = config('Confirm email', {
  tag: 'access',
  roles: ['guest', 'customer', 'contractor'],
  schema: {
    body: {
      type: 'object',
      required: ['code'],
      properties: {
        code: { type: 'string' },
      },
      additionalProperties: false,
    },
    response: {
      200: { $ref: 'models#/properties/profile' },
      401: { $ref: 'errors#/properties/simple' },
    },
  },
});

export const handler: Handler<ConfirmEmailPost> = async (request, reply) => {
  const { code } = request.body;

  const user = await findUserByEmailConfirmationCode(code);

  const isChangeEmailAction = !!user?.private.email.temporaryEmail?.value;

  if (
    !user ||
    !user.private.email.confirm ||
    (user.contacts.email.isVerified && !isChangeEmailAction)
  ) {
    return reply.code(401).send({ error: ERRORS.CONFIRMATION_CODE_NOT_VALID });
  }

  if (dayjs().isAfter(dayjs(user.private.email.confirm.expiredAt))) {
    return reply.code(401).send({ error: ERRORS.CONFIRMATION_CODE_EXPIRED });
  }

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const newUser = (await confirmUserEmail(user._id))!;

  if (!request.isAuthenticated()) {
    await signIn(request, newUser);
  }

  return userToProfile(newUser as User);
};
