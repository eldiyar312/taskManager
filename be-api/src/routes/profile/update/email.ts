import { NOTIFICATIONS } from 'shared';

import { Handler, ProfileEmailPut } from 'src/_generated';
import { User } from 'src/models/user';
import { notifications } from 'src/services';
import { userToProfile } from 'src/storage/profile';
import { updateUserById } from 'src/storage/users';
import { generateRandomStringBasedOnString } from 'src/utils/auth';
import { rollNowDateForward } from 'src/utils/dates';
import { config } from 'src/utils/navigation';

export const options = config('Update profile email', {
  tag: 'access',
  roles: ['customer', 'contractor'],
  schema: {
    body: {
      type: 'object',
      required: ['email'],
      properties: {
        email: { type: 'string', format: 'email' },
      },
      additionalProperties: false,
    },
    response: {
      200: { $ref: 'models#/properties/profile' },
      400: { $ref: 'errors#/properties/simple' },
    },
  },
});

export const handler: Handler<ProfileEmailPut> = async (request) => {
  const { email } = request.body;

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const user = request.user!;

  const isNewEmail = email && user.contacts.email.value !== email;

  if (!isNewEmail) {
    return userToProfile(user as User);
  }

  const confirmCode = generateRandomStringBasedOnString(email);

  const newUser = await updateUserById(user._id, {
    temporaryEmail: email,
    confirm: {
      code: confirmCode,
      expiredAt: rollNowDateForward(),
    },
  });

  await notifications.email(NOTIFICATIONS.CHANGING_EMAIL, {
    locale: user.regional.locale,
    code: confirmCode,
    email,
  });

  return userToProfile(newUser as User);
};
