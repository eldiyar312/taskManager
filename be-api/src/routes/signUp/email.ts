import { ERRORS, NOTIFICATIONS } from 'shared';

import { Handler, SignUpEmailPost } from 'src/_generated';
import { signIn } from 'src/middlewares/passport';
import { User } from 'src/models/user';
import { notifications } from 'src/services';
import { userToProfile } from 'src/storage/profile';
import { createUser, findUserByEmail } from 'src/storage/users';
import {
  createPasswordHash,
  generateRandomString,
  generateRandomStringBasedOnString,
  verifyPassword,
} from 'src/utils/auth';
import { rollNowDateForward } from 'src/utils/dates';
import { config } from 'src/utils/navigation';

export const options = config('Sign Up with email and password', {
  tag: 'access',
  roles: ['guest'],
  schema: {
    body: {
      type: 'object',
      required: ['email', 'isAgreeTerms', 'locale', 'password', 'role'],
      properties: {
        email: { type: 'string', format: 'email' },
        isAgreeTerms: { type: 'boolean' },
        locale: { $ref: 'enums#/properties/locale' },
        password: { type: 'string', minLength: 8 },
        role: { $ref: 'enums#/properties/role' },
      },
      additionalProperties: false,
    },
    response: {
      200: { $ref: 'models#/properties/profile' },
      401: { $ref: 'errors#/properties/simple' },
      451: { $ref: 'errors#/properties/simple' },
    },
  },
});

export const handler: Handler<SignUpEmailPost> = async (request, reply) => {
  const { email, isAgreeTerms, locale, password, role } = request.body;

  if (!isAgreeTerms) {
    return reply.code(451).send({ error: ERRORS.NEED_TO_ACCEPT_TERMS });
  }

  const user = await findUserByEmail(email);

  if (user) {
    // If a user registers with an email and password that are already
    // in the database, then we can authorize him.
    if (verifyPassword(password, user.private.password)) {
      await signIn(request, user);

      return userToProfile(user as User);
    }

    return reply.code(401).send({ error: ERRORS.EMAIL_ALREADY_EXISTS });
  }

  const salt = generateRandomString();
  const hash = createPasswordHash(password, salt);

  const code = generateRandomStringBasedOnString(email);

  const newUser = await createUser({
    agreements: { terms: isAgreeTerms },
    confirm: { code, expiredAt: rollNowDateForward() },
    personal: { name: 'Test User' },
    email,
    hash,
    salt,
    locale,
    role,
  });

  await notifications.email(NOTIFICATIONS.SIGN_UP_EMAIL, {
    locale,
    code,
    email,
  });

  await signIn(request, newUser);

  return userToProfile(newUser as User);
};
