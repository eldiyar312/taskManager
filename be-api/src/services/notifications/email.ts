import { CONTACTS, Locales, MODALS, NOTIFICATIONS, PAGES } from 'shared';

import { MsgProps, msg } from 'src/i18n/msg';
import { EmailTemplate, send } from 'src/integrations/sendgrid';
import { UserConfirmation } from 'src/models/user';
import { Email, Link } from 'src/types';
import { makeWebUrl } from 'src/utils/navigation';

type Tasks = keyof Task;
type TaskToHandler = {
  [K in Tasks]: (params: Task[K]) => EmailTemplate;
};

type CommonParams = {
  locale: Locales;
  email: Email;
  attachment?: Link[];
};

type Task = {
  [NOTIFICATIONS.CHANGING_EMAIL]: CommonParams & {
    code: UserConfirmation['code'];
  };
  [NOTIFICATIONS.FORGOT_PASSWORD]: CommonParams & {
    code: UserConfirmation['code'];
  };
  [NOTIFICATIONS.FORGOT_PASSWORD_CHANGE]: CommonParams;
  [NOTIFICATIONS.SIGN_UP_EMAIL]: CommonParams & {
    code: UserConfirmation['code'];
  };
  [NOTIFICATIONS.CHANGED_PASSWORD]: CommonParams;
};

type EmailProps = CommonParams & {
  subject: MsgProps;
  text: MsgProps[];
  html: MsgProps[];
};

const getFrom = (
  email: (typeof CONTACTS)['EMAILS'][keyof (typeof CONTACTS)['EMAILS']]
) => `Template <${email}>`;

const renderAttachment = (
  attachment: CommonParams['attachment'],
  mode: 'text' | 'html',
  locale: CommonParams['locale']
): string => {
  if (!attachment) {
    return '';
  }

  const text =
    mode === 'text'
      ? '\n' + attachment.join('\n')
      : attachment.map((link) => `<p><img src="${link}" /></p>`);

  const rendered = msg(locale, {
    id: 'services.notifications.email.attachments',
    values: { text },
  });

  return mode === 'text' ? rendered : `<p>${rendered}</p>`;
};

const makeEmail = ({
  locale,
  email,
  subject,
  text,
  html,
  attachment,
}: EmailProps): EmailTemplate => ({
  from: getFrom(CONTACTS.EMAILS.NOREPLY),
  to: email,
  subject: msg(locale, subject),
  text: `${text.map((line) => msg(locale, line)).join('\n')}

    ${msg(locale, {
      id: 'services.notifications.email.regards',
      values: { br: '\n' },
    })}

    ${renderAttachment(attachment, 'text', locale)}
    `.trim(),
  html: `${html.map((line) => `<p>${msg(locale, line)}</p>`).join('')}

    <p>${msg(locale, {
      id: 'services.notifications.email.regards',
      values: { br: '<br />' },
    })}</p>

    ${renderAttachment(attachment, 'html', locale)}
    `.trim(),
  attachment,
});

export const taskToHandler: TaskToHandler = {
  CHANGING_EMAIL: ({ locale, email, code }) => {
    const link = makeWebUrl([
      { scheme: PAGES.ROOT },
      { scheme: MODALS.CONFIRM_EMAIL, params: { code } },
    ]);

    return makeEmail({
      locale,
      email,
      subject: { id: 'services.notifications.email.changeEmail.title' },
      text: [
        { id: 'services.notifications.email.changeEmail.text1' },
        {
          id: 'services.notifications.email.changeEmail.text2.0',
          values: { link },
        },
      ],
      html: [
        { id: 'services.notifications.email.changeEmail.text1' },
        {
          id: 'services.notifications.email.changeEmail.text2.1',
          values: {
            link: `<a href="${link}">${msg(locale, {
              id: 'services.notifications.email.changeEmail.text2.2',
            })}</a>`,
          },
        },
      ],
    });
  },
  CHANGED_PASSWORD: ({ locale, email }) =>
    makeEmail({
      locale,
      email,
      subject: { id: 'services.notifications.email.changePassword.title' },
      text: [
        { id: 'services.notifications.email.changePassword.text1' },
        {
          id: 'services.notifications.email.changePassword.text2',
          values: { email: CONTACTS.EMAILS.SUPPORT },
        },
        { id: 'services.notifications.email.changePassword.text3' },
      ],
      html: [
        { id: 'services.notifications.email.changePassword.text1' },
        {
          id: 'services.notifications.email.changePassword.text2',
          values: { email: CONTACTS.EMAILS.SUPPORT },
        },
        { id: 'services.notifications.email.changePassword.text3' },
      ],
    }),
  FORGOT_PASSWORD: ({ locale, email, code }) => {
    const link = makeWebUrl([
      { scheme: PAGES.ROOT },
      { scheme: MODALS.FORGOT_PASSWORD_CHANGE, params: { code } },
    ]);

    return makeEmail({
      locale,
      email,
      subject: { id: 'services.notifications.email.forgotPassword.title' },
      text: [
        { id: 'services.notifications.email.forgotPassword.text1' },
        {
          id: 'services.notifications.email.forgotPassword.text2.0',
          values: { link },
        },
        { id: 'services.notifications.email.forgotPassword.text3' },
      ],
      html: [
        { id: 'services.notifications.email.forgotPassword.text1' },
        {
          id: 'services.notifications.email.forgotPassword.text2.1',
          values: {
            link: `<a href="${link}">${msg(locale, {
              id: 'services.notifications.email.forgotPassword.text2.2',
            })}</a>`,
          },
        },
        { id: 'services.notifications.email.forgotPassword.text3' },
      ],
    });
  },
  FORGOT_PASSWORD_CHANGE: ({ locale, email }) =>
    makeEmail({
      locale,
      email,
      subject: {
        id: 'services.notifications.email.forgotPasswordChange.title',
      },
      text: [
        { id: 'services.notifications.email.forgotPasswordChange.text1' },
        {
          id: 'services.notifications.email.forgotPasswordChange.text2',
          values: { email: CONTACTS.EMAILS.SUPPORT },
        },
        { id: 'services.notifications.email.forgotPasswordChange.text3' },
      ],
      html: [
        { id: 'services.notifications.email.forgotPasswordChange.text1' },
        {
          id: 'services.notifications.email.forgotPasswordChange.text2',
          values: {
            email: `<a href="mailto:${CONTACTS.EMAILS.SUPPORT}">${CONTACTS.EMAILS.SUPPORT}</a>`,
          },
        },
        { id: 'services.notifications.email.forgotPasswordChange.text3' },
      ],
    }),
  SIGN_UP_EMAIL: ({ locale, email, code }) => {
    const link = makeWebUrl([
      { scheme: PAGES.ROOT },
      { scheme: MODALS.CONFIRM_EMAIL, params: { code } },
    ]);

    return makeEmail({
      locale,
      email,
      subject: { id: 'services.notifications.email.signUpEmail.title' },
      text: [
        { id: 'services.notifications.email.signUpEmail.text1' },
        {
          id: 'services.notifications.email.signUpEmail.text2.0',
          values: { link },
        },
      ],
      html: [
        { id: 'services.notifications.email.signUpEmail.text1' },
        {
          id: 'services.notifications.email.signUpEmail.text2.1',
          values: {
            link: `<a href="${link}">${msg(locale, {
              id: 'services.notifications.email.signUpEmail.text2.2',
            })}</a>`,
          },
        },
      ],
    });
  },
};

export const email = async <T extends Tasks>(
  name: T,
  params: Task[T]
): Promise<{}[]> => send(taskToHandler[name](params));
