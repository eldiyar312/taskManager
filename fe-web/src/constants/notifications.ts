import { ERRORS } from 'shared';

import { DictionaryKey } from 'src/i18n';

export const errorToDictionaryKey: Record<ERRORS, DictionaryKey> = {
  [ERRORS.ALREADY_LOGGED_IN]: 'error.login.alreadyLoggedIn',
  [ERRORS.BAD_ENDPOINT]: 'error.endpoint.notValid',
  [ERRORS.CONFIRMATION_CODE_EXPIRED]: 'error.confirmationCode.expired',
  [ERRORS.CONFIRMATION_CODE_NOT_VALID]: 'error.confirmationCode.notValid',
  [ERRORS.EMAIL_ALREADY_EXISTS]: 'error.email.alreadyExists',
  [ERRORS.INCORRECT_PASSWORD]: 'error.password.notValid',
  [ERRORS.INVALID_EMAIL]: 'error.email.notValid',
  [ERRORS.INVALID_EMAIL_OR_PASSWORD]: 'error.emailOrPassword.notValid',
  [ERRORS.INVALID_ID]: 'error.endpoint.invalidId',
  [ERRORS.NEED_TO_ACCEPT_TERMS]: 'error.terms.needToAccept',
  [ERRORS.NEED_TO_LOGIN]: 'error.login.need',
  [ERRORS.PERMISSION_DENIED]: 'error.permissionDenied',
  [ERRORS.USER_NOT_FOUND]: 'error.user.notFound',
  [ERRORS.LACK_OF_FUNDS]: 'error.user.lackOfFunds',
  [ERRORS.INVALID_DATE]: 'error.user.invalidDate',
  [ERRORS.INVALID_MEDIA_TYPE]: 'error.media.invalidType',
  [ERRORS.INVALID_FORM_DATA]: 'error.media.invalidFormData',
  [ERRORS.MEDIA_SIZE_TOO_LARGE]: 'error.media.sizeTooLarge',
  [ERRORS.INSUFFICIENT_STORAGE]: 'error.media.insufficientStorage',
};
