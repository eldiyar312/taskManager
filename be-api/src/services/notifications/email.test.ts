import { LOCALES } from 'shared';

import { taskToHandler } from 'src/services/notifications/email';

const common = {
  email: 'user@example.com',
  code: 'xxxyyy',
};

const handlers = Object.entries(taskToHandler);

describe('Notifications: email', () => {
  handlers.forEach(([event, handler]) => {
    describe(`Event: ${event}`, () => {
      LOCALES.forEach((locale) => {
        test(`Notification with '${locale}' locale works correctly`, () => {
          expect(handler({ locale, ...common })).toMatchSnapshot();
        });
      });
    });
  });
});
