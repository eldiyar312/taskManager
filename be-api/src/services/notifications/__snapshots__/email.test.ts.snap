// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications: email Event: CHANGED_PASSWORD Notification with 'ru' locale works correctly 1`] = `
{
  "attachment": undefined,
  "from": "Template <noreply@todo.com>",
  "html": "<p>Вы получили это письмо, так как изменили пароль для входа в Template.</p><p>Если это были не вы, свяжитесь со службой поддержки: support@todo.com.</p><p>Если это были вы, ничего делать не нужно.</p>

    <p>С уважением,<br />Template</p>",
  "subject": "Ваш пароль был изменён",
  "text": "Вы получили это письмо, так как изменили пароль для входа в Template.
Если это были не вы, свяжитесь со службой поддержки: support@todo.com.
Если это были вы, ничего делать не нужно.

    С уважением,
Template",
  "to": "user@example.com",
}
`;

exports[`Notifications: email Event: CHANGING_EMAIL Notification with 'ru' locale works correctly 1`] = `
{
  "attachment": undefined,
  "from": "Template <noreply@todo.com>",
  "html": "<p>Была предпринята попытка изменить адрес электронной почты в вашей учетной записи!</p><p>Чтобы завершить процесс изменения адреса электронной почты, <a href="http://localhost:3000/~/confirm/email/xxxyyy">нажмите сюда</a>.</p>

    <p>С уважением,<br />Template</p>",
  "subject": "Пожалуйста, подтвердите новый адрес электронной почты",
  "text": "Была предпринята попытка изменить адрес электронной почты в вашей учетной записи!
Чтобы завершить процесс изменения адреса электронной почты, перейдите по ссылке: http://localhost:3000/~/confirm/email/xxxyyy

    С уважением,
Template",
  "to": "user@example.com",
}
`;

exports[`Notifications: email Event: FORGOT_PASSWORD Notification with 'ru' locale works correctly 1`] = `
{
  "attachment": undefined,
  "from": "Template <noreply@todo.com>",
  "html": "<p>Вы получили это письмо, так как запросили восстановление пароля для входа в Template</p><p><a href="http://localhost:3000/~/forgot/password/change/xxxyyy">Нажмите сюда</a>, чтобы задать себе новый пароль.</p><p>Если это были не вы, просто проигнорируйте данное письмо.</p>

    <p>С уважением,<br />Template</p>",
  "subject": "Восстановление пароля",
  "text": "Вы получили это письмо, так как запросили восстановление пароля для входа в Template
Перейдите по ссылке, чтобы задать себе новый пароль: http://localhost:3000/~/forgot/password/change/xxxyyy
Если это были не вы, просто проигнорируйте данное письмо.

    С уважением,
Template",
  "to": "user@example.com",
}
`;

exports[`Notifications: email Event: FORGOT_PASSWORD_CHANGE Notification with 'ru' locale works correctly 1`] = `
{
  "attachment": undefined,
  "from": "Template <noreply@todo.com>",
  "html": "<p>Вы получили это письмо, так как прошли процедуру восстановления и смены пароля для входа в Template</p><p>Если это были не вы, свяжитесь со службой поддержки: <a href="mailto:support@todo.com">support@todo.com</a></p><p>Если это были вы, ничего делать не нужно.</p>

    <p>С уважением,<br />Template</p>",
  "subject": "Ваш пароль был изменен",
  "text": "Вы получили это письмо, так как прошли процедуру восстановления и смены пароля для входа в Template
Если это были не вы, свяжитесь со службой поддержки: support@todo.com
Если это были вы, ничего делать не нужно.

    С уважением,
Template",
  "to": "user@example.com",
}
`;

exports[`Notifications: email Event: SIGN_UP_EMAIL Notification with 'ru' locale works correctly 1`] = `
{
  "attachment": undefined,
  "from": "Template <noreply@todo.com>",
  "html": "<p>Благодарим вас за регистрацию!</p><p>Чтобы завершить процесс регистрации, <a href="http://localhost:3000/~/confirm/email/xxxyyy">нажмите сюда</a>.</p>

    <p>С уважением,<br />Template</p>",
  "subject": "Пожалуйста, подтвердите свой адрес электронной почты",
  "text": "Благодарим вас за регистрацию!
Чтобы завершить процесс регистрации, перейдите по ссылке: http://localhost:3000/~/confirm/email/xxxyyy

    С уважением,
Template",
  "to": "user@example.com",
}
`;
